### DISCUSSION TITLE
Next.js App Router - how to organize server & client files in different folders?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [vercel/next.js#52766] [Help] [answered]
### Summary

i had separate `client` & `server` folders:

```bash
.
├── client
│   ├── components
│   │   ├── AuthenticationForm
│   │   │   └── index.tsx
│   │   ├── Footer
│   │   │   └── index.tsx
│   │   ├── Layout
│   │   │   └── index.tsx
│   │   ├── Navbar
│   │   │   └── index.tsx
│   │   └── UpgradeButton
│   │       └── index.tsx
│   └── graphql
│       ├── client.ts
│       ├── createProject.graphql
│       ├── createStripeCheckoutBillingPortalUrl.graphql
│       ├── createStripeCheckoutSession.graphql
│       ├── getCurrentUser.graphql
│       ├── getCurrentUserWithProjects.graphql
│       ├── getProject.graphql
│       ├── getProjectWithUsers.graphql
│       ├── inviteToProject.graphql
│       ├── removeUserFromProject.graphql
│       └── updateUser.graphql
├── pages
│   ├── _app.tsx
│   ├── api
│   │   ├── auth
│   │   │   ├── logout.ts
│   │   │   └── magiclink
│   │   │       ├── callback.ts
│   │   │       └── index.ts
│   │   ├── index.ts
│   │   ├── invitations
│   │   │   └── accept.ts
│   │   └── stripe
│   │       └── webhooks.ts
│   ├── app
│   │   ├── [slug]
│   │   │   ├── index.tsx
│   │   │   └── settings.tsx
│   │   ├── index.tsx
│   │   └── settings.tsx
│   ├── check-mailbox.tsx
│   ├── get-started.tsx
│   ├── index.tsx
│   └── login.tsx
├── server
│   ├── api-route.ts
│   ├── db
│   │   ├── migrations
│   │   │   ├── 20210218175246_initial_schema
│   │   │   │   └── migration.sql
│   │   │   ├── 20220205085812_prisma_upgrade
│   │   │   │   └── migration.sql
│   │   │   └── migration_lock.toml
│   │   ├── prisma.ts
│   │   └── schema.prisma
│   ├── get-project-paid-plan.ts
│   ├── get-request-origin.ts
│   ├── graphql
│   │   ├── Project
│   │   │   └── index.ts
│   │   ├── User
│   │   │   ├── index.ts
│   │   │   └── test.ts
│   │   ├── schema.graphql
│   │   └── schema.ts
│   ├── invitations
│   │   └── token.ts
│   ├── passport
│   │   ├── index.ts
│   │   └── magicLink.ts
│   ├── send-email.ts
│   ├── stripe
│   │   ├── index.ts
│   │   └── plans.ts
│   └── trust-proxy-middleware.ts
└── test
    ├── jest-setup.ts
    ├── request.ts
    └── seed
        ├── data.ts
        └── index.ts

29 directories, 55 files
```

but how do i do that in app directory?

i want to separate so i put server stuff in different directory & client in different so i can import easily without thinking? currently, i get confused while importing stripe as it has 2 different ways to import (client & server)

**Comment by deadcoder0904** [comment]
this one looks okay to me but lemme know if anyone can improve over this:

	.
	├── public
	└── src
			└── app
					├── (client)
					│   ├── components
					│   │   ├── homepage
					│   │   └── resend
					│   ├── lib
					│   └── styles
					├── (pages)
					│   ├── admin
					│   ├── dashboard
					│   ├── login
					│   └── verify-email
					└── (server)
							├── api
							│   ├── download-pdf
							│   ├── stripe-webhook
							│   └── upload-pdf
							├── auth
							├── db
							│   └── migrations
							│       └── meta
							└── lib

	24 directories

**Reply by icyJoseph** [reply, marked as answer]
You have to use Private folders for the (client) subfolder.

- https://nextjs.org/docs/app/building-your-application/routing/colocation#private-folders

This page has all of the info you need though: https://nextjs.org/docs/app/building-your-application/routing/colocation

The way you suggest here, as soon as there's a `page.tsx` in the (client) subfolder, that'll be a reachable route.

I think you are also creating a meta layer here, because you'll have (client) subfolder stuff that'll be opted in to server components, unless there's a 'use client' directive at the top, or imported by a use client sub-tree.

I think `src/components` is enough, and keep your server stuff in some `src/lib` and for example stripe, `src/lib/stripe/client.ts` and `src/lib/stripe/server.ts` to keep the environment, browser or node, usage clear.

**Reply by deadcoder0904** [reply]
thanks @icyJoseph i like your example.

i have to change my whole mental model i guess to how i did stuff previously.

the client & server components don't yet make sense. i guess i have to practice it a few times before i get this rsc stuff.

**Reply by icyJoseph** [reply]
Yeah, in all honesty, there's a bunch of concepts and conventions that we have to get used to. Some are intuitive, others aren't as much, like interceptors and parallel routes, as well as having a clear map of when a component is opted-in to Client Component, because it is part of a client tree and not because it has a `use client` directive, etc.
